File: fetch_all_candles.py
# fetch_all_candles.py
import requests
import time
import json

API_BASE = "https://final-ocai.onrender.com"

def get_all_companies():
    response = requests.get(f"{API_BASE}/api/companies/all")
    return response.json()["data"]

def get_candles(symbol, period="1y"):
    try:
        response = requests.get(f"{API_BASE}/api/candles/{symbol}?period={period}")
        return response.json()
    except:
        return None

def fetch_all_candles(period="1y"):
    print("Fetching all companies...")
    companies = get_all_companies()
    print(f"Found {len(companies)} companies")
    
    results = []
    for i, company in enumerate(companies):
        print(f"[{i+1}/{len(companies)}] Fetching {company['symbol']}...")
        candle_data = get_candles(company['symbol'], period)
        
        if candle_data and candle_data.get("success"):
            results.append({
                "symbol": company['symbol'],
                "name": company['name'],
                "count": candle_data["count"],
                "data": candle_data["data"]
            })
            print(f"  ✅ {candle_data['count']} candles")
        else:
            print(f"  ❌ No data")
        
        time.sleep(0.5)  # Rate limiting
    
    # Save to file
    output = {
        "generated_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "period": period,
        "total_companies": len(companies),
        "data": results
    }
    
    with open("all_candles.json", "w") as f:
        json.dump(output, f, indent=2)
    
    print(f"\n✅ Saved to all_candles.json")

if __name__ == "__main__":
    period = input("Enter period (1m/3m/6m/1y/3y/5y): ") or "1y"
    fetch_all_candles(period)

File: scripts/fetch-all-candles.js
// scripts/fetch-all-candles.js
const axios = require('axios');
const fs = require('fs');

const API_BASE = 'https://final-ocai.onrender.com';
const OUTPUT_FILE = 'all-candles-data.json';

async function fetchAllCompanies() {
  try {
    const response = await axios.get(`${API_BASE}/api/companies/all`);
    return response.data.data;
  } catch (error) {
    console.error('Error fetching companies:', error.message);
    return [];
  }
}

async function fetchCandles(symbol, period = '1y') {
  try {
    const response = await axios.get(`${API_BASE}/api/candles/${symbol}?period=${period}`);
    return response.data;
  } catch (error) {
    console.error(`Error fetching candles for ${symbol}:`, error.message);
    return null;
  }
}

async function fetchAllCandles(period = '1y', batchSize = 5) {
  console.log('🚀 Fetching all companies...');
  const companies = await fetchAllCompanies();
  
  if (companies.length === 0) {
    console.log('❌ No companies found');
    return;
  }
  
  console.log(`📊 Found ${companies.length} companies`);
  console.log(`📈 Fetching ${period} of candle data for each...\n`);
  
  const results = [];
  let successCount = 0;
  let failCount = 0;
  
  for (let i = 0; i < companies.length; i++) {
    const company = companies[i];
    console.log(`[${i + 1}/${companies.length}] Fetching ${company.symbol}...`);
    
    const candleData = await fetchCandles(company.symbol, period);
    
    if (candleData && candleData.success && candleData.count > 0) {
      results.push({
        symbol: company.symbol,
        name: company.name,
        period: period,
        count: candleData.count,
        data: candleData.data
      });
      successCount++;
      console.log(`  ✅ ${candleData.count} candles retrieved`);
    } else {
      failCount++;
      console.log(`  ❌ No data available`);
    }
    
    // Wait between requests to avoid rate limiting
    await new Promise(resolve => setTimeout(resolve, 500));
  }
  
  // Save results to file
  const output = {
    generated_at: new Date().toISOString(),
    period: period,
    total_companies: companies.length,
    success_count: successCount,
    fail_count: failCount,
    data: results
  };
  
  fs.writeFileSync(OUTPUT_FILE, JSON.stringify(output, null, 2));
  console.log(`\n✅ Saved to ${OUTPUT_FILE}`);
  console.log(`📊 Summary: ${successCount} successful, ${failCount} failed`);
  
  return output;
}

// Run with command line argument for period
const period = process.argv[2] || '1y';
fetchAllCandles(period).catch(console.error);